Panic "Cannot add pos to new symbol" when multiple struct fields share an anonymous struct as their type (concise syntax)

**Summary**

A dependency of ours is crashing when we run `scip-go`:

```
panic: Cannot add pos to new symbol: symbol:"scip-go gomod github.com/rivo/tview 0d02bb78244d `github.com/rivo/tview`/TextArea#cursor.row."  documentation:"```go\nstruct field row int\n```" symbol:"scip-go gomod github.com/rivo/tview 0d02bb78244d `github.com/rivo/tview`/TextArea#cursor.selectionStart.row."  documentation:"```go\nstruct field row int\n```"
```

Click here to view full log output of `time scip-go`

```
➜ time scip-go
scip-go
Resolving module name
  Go standard library version:  go1.18
  Resolved module name       :  github.com/rivo/tview

Configuration:
  Skip Implementations: false
  Skip Test           : false
Loading Packages
Visiting Packages
panic: Cannot add pos to new symbol: symbol:"scip-go gomod github.com/rivo/tview 0d02bb78244d `github.com/rivo/tview`/TextArea#cursor.row."  documentation:"```go\nstruct field row int\n```" symbol:"scip-go gomod github.com/rivo/tview 0d02bb78244d `github.com/rivo/tview`/TextArea#cursor.selectionStart.row."  documentation:"```go\nstruct field row int\n```"

goroutine 1793 [running]:
github.com/sourcegraph/scip-go/internal/lookup.(*Package).Set(0x1400ad46320, 0x9dcd19, 0x1400ea19400)
	/Users/qaisjp/go/pkg/mod/github.com/sourcegraph/scip-go@v0.1.12/internal/lookup/lookup.go:51 +0xf8
github.com/sourcegraph/scip-go/internal/document.(*Document).SetNewSymbolForPos(0x1400e9dd110, {0x1400ea00780, 0x6c}, {0x0?, 0x0?}, 0x1401530f6e0?, 0x9dcd19)
	/Users/qaisjp/go/pkg/mod/github.com/sourcegraph/scip-go@v0.1.12/internal/document/document.go:103 +0x378
github.com/sourcegraph/scip-go/internal/document.(*Document).SetNewSymbol(...)
	/Users/qaisjp/go/pkg/mod/github.com/sourcegraph/scip-go@v0.1.12/internal/document/document.go:69
github.com/sourcegraph/scip-go/internal/visitors.typeVisitor.Visit({0x1400e9dd110, 0x140004eb980, 0x1400ea02330, 0x14015311880, 0x0}, {0x10094ec90, 0x140153113c0})
	/Users/qaisjp/go/pkg/mod/github.com/sourcegraph/scip-go@v0.1.12/internal/visitors/visitor_type.go:117 +0x764
go/ast.Walk({0x10094c400?, 0x1400ea05e30?}, {0x10094ec90, 0x140153113c0})
	/opt/homebrew/Cellar/go/1.22.0/libexec/src/go/ast/walk.go:51 +0x44
go/ast.Walk({0x10094c400?, 0x1400ea05dd0?}, {0x10094ece0, 0x14015303260})
	/opt/homebrew/Cellar/go/1.22.0/libexec/src/go/ast/walk.go:85 +0x1e68
go/ast.Walk({0x10094c400?, 0x1400ea05d70?}, {0x10094ef60, 0x1401532a708})
	/opt/homebrew/Cellar/go/1.22.0/libexec/src/go/ast/walk.go:168 +0xdc8
github.com/sourcegraph/scip-go/internal/visitors.typeVisitor.Visit({0x1400e9dd110, 0x140004eb980, 0x1400ea02330, 0x14015311880, 0x0}, {0x10094ec90, 0x14015311440})
	/Users/qaisjp/go/pkg/mod/github.com/sourcegraph/scip-go@v0.1.12/internal/visitors/visitor_type.go:149 +0xe84
go/ast.Walk({0x10094c400?, 0x1400ea04990?}, {0x10094ec90, 0x14015311440})
	/opt/homebrew/Cellar/go/1.22.0/libexec/src/go/ast/walk.go:51 +0x44
go/ast.Walk({0x10094c400?, 0x1400ea04930?}, {0x10094ece0, 0x140153033e0})
	/opt/homebrew/Cellar/go/1.22.0/libexec/src/go/ast/walk.go:85 +0x1e68
go/ast.Walk({0x10094c400?, 0x1400ea048d0?}, {0x10094ef60, 0x1401532aaf8})
	/opt/homebrew/Cellar/go/1.22.0/libexec/src/go/ast/walk.go:168 +0xdc8
github.com/sourcegraph/scip-go/internal/visitors.typeVisitor.Visit({0x1400e9dd110, 0x140004eb980, 0x1400ea02330, 0x14015311880, 0x0}, {0x10094f370, 0x14015310c40})
	/Users/qaisjp/go/pkg/mod/github.com/sourcegraph/scip-go@v0.1.12/internal/visitors/visitor_type.go:94 +0x488
go/ast.Walk({0x10094c400?, 0x1400ea04810?}, {0x10094f370, 0x14015310c40})
	/opt/homebrew/Cellar/go/1.22.0/libexec/src/go/ast/walk.go:51 +0x44
github.com/sourcegraph/scip-go/internal/visitors.typeVisitor.Visit({0x1400e9dd110, 0x140004eb980, 0x1400ea02330, 0x0, 0x0}, {0x10094f3c0, 0x14015311880})
	/Users/qaisjp/go/pkg/mod/github.com/sourcegraph/scip-go@v0.1.12/internal/visitors/visitor_type.go:63 +0x4f0
go/ast.Walk({0x10094c400?, 0x1400e9dd1a0?}, {0x10094f3c0, 0x14015311880})
	/opt/homebrew/Cellar/go/1.22.0/libexec/src/go/ast/walk.go:51 +0x44
go/ast.walkDeclList(...)
	/opt/homebrew/Cellar/go/1.22.0/libexec/src/go/ast/walk.go:38
go/ast.Walk({0x10094c400?, 0x1400e9dd140?}, {0x10094e908, 0x14012e6cf00})
	/opt/homebrew/Cellar/go/1.22.0/libexec/src/go/ast/walk.go:366 +0x2e14
github.com/sourcegraph/scip-go/internal/visitors.visitTypesInFile(0x1400e9dd110, 0x140004eb980, 0x14012e6cf00)
	/Users/qaisjp/go/pkg/mod/github.com/sourcegraph/scip-go@v0.1.12/internal/visitors/visitor_type.go:19 +0x11c
github.com/sourcegraph/scip-go/internal/visitors.visitSyntax(0x140004eb980, 0x1400ad46320, 0x14012e6cf00, {0x1400003ed3b, 0xb})
	/Users/qaisjp/go/pkg/mod/github.com/sourcegraph/scip-go@v0.1.12/internal/visitors/visitors.go:45 +0xd4
github.com/sourcegraph/scip-go/internal/visitors.VisitPackageSyntax({0x1400003b8c0, 0x1a}, 0x140004eb980, 0x1400bace2a0, 0x1400a90adf8)
	/Users/qaisjp/go/pkg/mod/github.com/sourcegraph/scip-go@v0.1.12/internal/visitors/visitors.go:30 +0x138
github.com/sourcegraph/scip-go/internal/index.indexVisitPackages.func1()
	/Users/qaisjp/go/pkg/mod/github.com/sourcegraph/scip-go@v0.1.12/internal/index/scip.go:191 +0xf8
created by github.com/sourcegraph/scip-go/internal/index.indexVisitPackages in goroutine 1
	/Users/qaisjp/go/pkg/mod/github.com/sourcegraph/scip-go@v0.1.12/internal/index/scip.go:186 +0x220
scip-go  1.88s user 0.73s system 44% cpu 5.948 total
```

**Reproduction steps**

1. Clone [https://github.com/rivo/tview](https://github.com/rivo/tview) (HEAD as of writing is [https://github.com/rivo/tview/commit/0d02bb78244d2636d19f7d1455a796fd7f9b0330](https://github.com/rivo/tview/commit/0d02bb78244d2636d19f7d1455a796fd7f9b0330))
2. Run `scip-go`

**Additional context**

The panic is happening here:

[https://github.com/sourcegraph/scip-go/blob/a5b080c35c9af6d3dfaa3dcdb209e2425630892d/internal/lookup/lookup.go#L47-L56](https://github.com/sourcegraph/scip-go/blob/a5b080c35c9af6d3dfaa3dcdb209e2425630892d/internal/lookup/lookup.go#L47-L56)

It's caused by [these lines](https://github.com/rivo/tview/blob/e22ce9588bb4e7b1206ce7b4fac5006a34c70365/textarea.go#L277-L292):

```go
 // The cursor always points to the next position where a new character would
 // be placed. The selection start is the same as cursor as long as there is
 // no selection. When there is one, the selection is between selectionStart
 // and cursor.
 cursor, selectionStart struct {
  // The row and column in screen space but relative to the start of the
  // text which may be outside the text area's box. The column value may
  // be larger than where the cursor actually is if the line the cursor
  // is on is shorter. The actualColumn is the position as it is seen on
  // screen. These three values may not be determined yet, in which case
  // the row is negative.
  row, column, actualColumn int
 
  // The textAreaSpan position with state for the actual next character.
  pos [3]int
 }
```

The following [patch](https://github.com/rivo/tview/commit/591ebbb71cba37fca67bb41c44dfc7324881bd2c) makes the error go away:

```diff
diff --git a/textarea.go b/textarea.go
index e6452e3c..116c7439 100644
--- a/textarea.go
+++ b/textarea.go
@@ -89,6 +89,19 @@ type textAreaUndoItem struct {
 	continuation                  bool   // If true, this item is a continuation of the previous undo item. It is handled together with all other undo items in the same continuation sequence.
 }
 
+type cursorOrSelectionStart struct {
+	// The row and column in screen space but relative to the start of the
+	// text which may be outside the text area's box. The column value may
+	// be larger than where the cursor actually is if the line the cursor
+	// is on is shorter. The actualColumn is the position as it is seen on
+	// screen. These three values may not be determined yet, in which case
+	// the row is negative.
+	row, column, actualColumn int
+
+	// The textAreaSpan position with state for the actual next character.
+	pos [3]int
+}
+
 // TextArea implements a simple text editor for multi-line text. Multi-color
 // text is not supported. Word-wrapping is enabled by default but can be turned
 // off or be changed to character-wrapping.
@@ -284,18 +297,8 @@ type TextArea struct {
 	// be placed. The selection start is the same as the cursor as long as there
 	// is no selection. When there is one, the selection is between
 	// selectionStart and cursor.
-	cursor, selectionStart struct {
-		// The row and column in screen space but relative to the start of the
-		// text which may be outside the text area's box. The column value may
-		// be larger than where the cursor actually is if the line the cursor
-		// is on is shorter. The actualColumn is the position as it is seen on
-		// screen. These three values may not be determined yet, in which case
-		// the row is negative.
-		row, column, actualColumn int
-
-		// The textAreaSpan position with state for the actual next character.
-		pos [3]int
-	}
+	cursor         cursorOrSelectionStart
+	selectionStart cursorOrSelectionStart
 
 	// The minimum width of text (if available) to be shown left of the cursor.
 	minCursorPrefix int
```

--- a/.vscode/launch.json
+++ b/.vscode/launch.json
@@ -1,0 +1,21 @@
+{
+  "version": "0.2.0",
+  "configurations": [
+    {
+      "name": "Run SCIP-GO",
+      "type": "go",
+      "request": "launch",
+      "mode": "auto",
+      "program": "cmd/scip-go",
+      "cwd": "${input:path}", 
+    }
+  ],
+  "inputs": [
+    {
+      "id": "path",
+      "description": "Please enter the path to the project to index",
+      "default": "",
+      "type": "promptString"
+    }
+  ]
+}

--- a/internal/lookup/lookup.go
+++ b/internal/lookup/lookup.go
@@ -45,16 +45,23 @@ func (p *Package) SymbolsForFile(file *token.File) []*scip.SymbolInformation {
 	return documentSymbols
 }
 
+var emittedLogLine = map[token.Pos]struct{}{}
+var emittedLogLineMu sync.Mutex
+
 func (p *Package) Set(pos token.Pos, symbol *scip.SymbolInformation) {
 	if original, ok := p.fields[pos]; ok {
 		if original != symbol {
-			// This is a temporary fix to allow overriding of symbols to handle a conflict
-			// when processing anonymous structs. Today, anonymous structs are named by the associated field name.
-			// But defining multiple fields with the same struct leads to this conflict. By allowing overriding,
-			// the last field name seen will be picked to be the type name for the struct.
-			// The right fix longer term is to change the way we name anonymous structs tracked here
-			// https://github.com/sourcegraph/scip-go/issues/95
-			output.Logf("[scip.lookup] Overriding original symbol %s with %s at %s", original.Symbol, symbol.Symbol, p.pkg.Fset.Position(pos))
+			// Workaround for handling symbol names when multiple
+			// fields are defined in the same statement with the same anonymous
+			// struct type. By ignoring this case, the last field name will
+			// be used for the type name.
+			// Ideal fix: https://github.com/sourcegraph/scip-go/issues/95
+			emittedLogLineMu.Lock()
+			if _, ok := emittedLogLine[pos]; !ok {
+				output.Logf("[scip.lookup] Overriding original symbol %s with %s at %v", original.Symbol, symbol.Symbol, p.pkg.Fset.Position(pos))
+				emittedLogLine[pos] = struct{}{}
+			}
+			emittedLogLineMu.Unlock()
 		}
 	}
 

--- a/internal/lookup/lookup.go
+++ b/internal/lookup/lookup.go
@@ -9,6 +9,7 @@ import (
 	"sync"
 
 	"github.com/sourcegraph/scip-go/internal/newtypes"
+	"github.com/sourcegraph/scip-go/internal/output"
 	"github.com/sourcegraph/scip-go/internal/symbols"
 	"github.com/sourcegraph/scip/bindings/go/scip"
 	"golang.org/x/tools/go/packages"
@@ -45,10 +46,15 @@ func (p *Package) SymbolsForFile(file *token.File) []*scip.SymbolInformation {
 }
 
 func (p *Package) Set(pos token.Pos, symbol *scip.SymbolInformation) {
-	// TODO: Could remove this once we are 100% confident we're not overlapping...
 	if original, ok := p.fields[pos]; ok {
 		if original != symbol {
-			panic(fmt.Sprintf("Cannot add pos to new symbol: %s %s", original, symbol))
+			// This is a temporary fix to allow overriding of symbols to handle a conflict
+			// when processing anonymous structs. Today, anonymous structs are named by the associated field name.
+			// But defining multiple fields with the same struct leads to this conflict. By allowing overriding,
+			// the last field name seen will be picked to be the type name for the struct.
+			// The right fix longer term is to change the way we name anonymous structs tracked here
+			// https://github.com/sourcegraph/scip-go/issues/95
+			output.Logf("[scip.lookup] Overriding original symbol %s with %s at %s", original.Symbol, symbol.Symbol, p.pkg.Fset.Position(pos))
 		}
 	}
 

--- a/internal/testdata/snapshots/input/testdata/anonymous_structs.go
+++ b/internal/testdata/snapshots/input/testdata/anonymous_structs.go
@@ -1,0 +1,47 @@
+package testdata
+
+import "fmt"
+
+type TypeContainingAnonymousStructs struct {
+	a, b struct {
+		x int
+		y string
+	}
+	c struct {
+		X int
+		Y string
+	}
+}
+
+func funcContainingAnonymousStructs() {
+	d := struct {
+		x int
+		y string
+	}{
+		x: 1,
+		y: "one",
+	}
+
+	var e struct {
+		x int
+		y string
+	}
+
+	e.x = 2
+	e.y = "two"
+
+	var f TypeContainingAnonymousStructs
+	f.a.x = 3
+	f.a.y = "three"
+	f.b.x = 4
+	f.b.y = "four"
+	f.c.X = 5
+	f.c.Y = "five"
+
+	fmt.Printf("> %s, %s\n", d.x, d.y)
+	fmt.Printf("> %s, %s\n", e.x, e.y)
+
+	fmt.Printf("> %s, %s\n", f.a.x, f.a.y)
+	fmt.Printf("> %s, %s\n", f.b.x, f.b.y)
+	fmt.Printf("> %s, %s\n", f.c.X, f.c.Y)
+}

--- a/internal/testdata/snapshots/output/testdata/anonymous_structs.go
+++ b/internal/testdata/snapshots/output/testdata/anonymous_structs.go
@@ -1,0 +1,137 @@
+  package testdata
+//        ^^^^^^^^ reference 0.1.test `sg/testdata`/
+  
+  import "fmt"
+//        ^^^ reference github.com/golang/go/src go1.22 fmt/
+  
+  type TypeContainingAnonymousStructs struct {
+//     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ definition 0.1.test `sg/testdata`/TypeContainingAnonymousStructs#
+//     documentation ```go
+//     documentation ```go
+   a, b struct {
+// ^ definition 0.1.test `sg/testdata`/TypeContainingAnonymousStructs#a.
+// documentation ```go
+//    ^ definition 0.1.test `sg/testdata`/TypeContainingAnonymousStructs#a.b.
+//    documentation ```go
+    x int
+//  ^ definition 0.1.test `sg/testdata`/TypeContainingAnonymousStructs#a.b.x.
+//  documentation ```go
+    y string
+//  ^ definition 0.1.test `sg/testdata`/TypeContainingAnonymousStructs#a.b.y.
+//  documentation ```go
+   }
+   c struct {
+// ^ definition 0.1.test `sg/testdata`/TypeContainingAnonymousStructs#c.
+// documentation ```go
+    X int
+//  ^ definition 0.1.test `sg/testdata`/TypeContainingAnonymousStructs#c.X.
+//  documentation ```go
+    Y string
+//  ^ definition 0.1.test `sg/testdata`/TypeContainingAnonymousStructs#c.Y.
+//  documentation ```go
+   }
+  }
+  
+  func funcContainingAnonymousStructs() {
+//     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ definition 0.1.test `sg/testdata`/funcContainingAnonymousStructs().
+//     documentation ```go
+   d := struct {
+// ^ definition local 0
+    x int
+//  ^ definition local 1
+    y string
+//  ^ definition local 2
+   }{
+    x: 1,
+//  ^ reference local 1
+    y: "one",
+//  ^ reference local 2
+   }
+  
+   var e struct {
+//     ^ definition local 3
+    x int
+//  ^ definition local 4
+    y string
+//  ^ definition local 5
+   }
+  
+   e.x = 2
+// ^ reference local 3
+//   ^ reference local 4
+   e.y = "two"
+// ^ reference local 3
+//   ^ reference local 5
+  
+   var f TypeContainingAnonymousStructs
+//     ^ definition local 6
+//       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ reference 0.1.test `sg/testdata`/TypeContainingAnonymousStructs#
+   f.a.x = 3
+// ^ reference local 6
+//   ^ reference 0.1.test `sg/testdata`/TypeContainingAnonymousStructs#a.
+//     ^ reference 0.1.test `sg/testdata`/TypeContainingAnonymousStructs#a.b.x.
+   f.a.y = "three"
+// ^ reference local 6
+//   ^ reference 0.1.test `sg/testdata`/TypeContainingAnonymousStructs#a.
+//     ^ reference 0.1.test `sg/testdata`/TypeContainingAnonymousStructs#a.b.y.
+   f.b.x = 4
+// ^ reference local 6
+//   ^ reference 0.1.test `sg/testdata`/TypeContainingAnonymousStructs#a.b.
+//     ^ reference 0.1.test `sg/testdata`/TypeContainingAnonymousStructs#a.b.x.
+   f.b.y = "four"
+// ^ reference local 6
+//   ^ reference 0.1.test `sg/testdata`/TypeContainingAnonymousStructs#a.b.
+//     ^ reference 0.1.test `sg/testdata`/TypeContainingAnonymousStructs#a.b.y.
+   f.c.X = 5
+// ^ reference local 6
+//   ^ reference 0.1.test `sg/testdata`/TypeContainingAnonymousStructs#c.
+//     ^ reference 0.1.test `sg/testdata`/TypeContainingAnonymousStructs#c.X.
+   f.c.Y = "five"
+// ^ reference local 6
+//   ^ reference 0.1.test `sg/testdata`/TypeContainingAnonymousStructs#c.
+//     ^ reference 0.1.test `sg/testdata`/TypeContainingAnonymousStructs#c.Y.
+  
+   fmt.Printf("> %s, %s\n", d.x, d.y)
+// ^^^ reference github.com/golang/go/src go1.22 fmt/
+//     ^^^^^^ reference github.com/golang/go/src go1.22 fmt/Printf().
+//                          ^ reference local 0
+//                            ^ reference local 1
+//                               ^ reference local 0
+//                                 ^ reference local 2
+   fmt.Printf("> %s, %s\n", e.x, e.y)
+// ^^^ reference github.com/golang/go/src go1.22 fmt/
+//     ^^^^^^ reference github.com/golang/go/src go1.22 fmt/Printf().
+//                          ^ reference local 3
+//                            ^ reference local 4
+//                               ^ reference local 3
+//                                 ^ reference local 5
+  
+   fmt.Printf("> %s, %s\n", f.a.x, f.a.y)
+// ^^^ reference github.com/golang/go/src go1.22 fmt/
+//     ^^^^^^ reference github.com/golang/go/src go1.22 fmt/Printf().
+//                          ^ reference local 6
+//                            ^ reference 0.1.test `sg/testdata`/TypeContainingAnonymousStructs#a.
+//                              ^ reference 0.1.test `sg/testdata`/TypeContainingAnonymousStructs#a.b.x.
+//                                 ^ reference local 6
+//                                   ^ reference 0.1.test `sg/testdata`/TypeContainingAnonymousStructs#a.
+//                                     ^ reference 0.1.test `sg/testdata`/TypeContainingAnonymousStructs#a.b.y.
+   fmt.Printf("> %s, %s\n", f.b.x, f.b.y)
+// ^^^ reference github.com/golang/go/src go1.22 fmt/
+//     ^^^^^^ reference github.com/golang/go/src go1.22 fmt/Printf().
+//                          ^ reference local 6
+//                            ^ reference 0.1.test `sg/testdata`/TypeContainingAnonymousStructs#a.b.
+//                              ^ reference 0.1.test `sg/testdata`/TypeContainingAnonymousStructs#a.b.x.
+//                                 ^ reference local 6
+//                                   ^ reference 0.1.test `sg/testdata`/TypeContainingAnonymousStructs#a.b.
+//                                     ^ reference 0.1.test `sg/testdata`/TypeContainingAnonymousStructs#a.b.y.
+   fmt.Printf("> %s, %s\n", f.c.X, f.c.Y)
+// ^^^ reference github.com/golang/go/src go1.22 fmt/
+//     ^^^^^^ reference github.com/golang/go/src go1.22 fmt/Printf().
+//                          ^ reference local 6
+//                            ^ reference 0.1.test `sg/testdata`/TypeContainingAnonymousStructs#c.
+//                              ^ reference 0.1.test `sg/testdata`/TypeContainingAnonymousStructs#c.X.
+//                                 ^ reference local 6
+//                                   ^ reference 0.1.test `sg/testdata`/TypeContainingAnonymousStructs#c.
+//                                     ^ reference 0.1.test `sg/testdata`/TypeContainingAnonymousStructs#c.Y.
+  }
+  

